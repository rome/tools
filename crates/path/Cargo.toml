[package]
name = "path"
version = "0.0.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
core = { path = "../core" }

[dev-dependencies]
