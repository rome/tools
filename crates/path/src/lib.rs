//! This module is responsible to manage paths inside Rome.
//! It is a small wrapper around [path::PathBuf] but it is also able to
//! give additional information around the the file that holds:
//! - the [FileHandlers] for the specific file
//! - shortcuts to open/write to the file
use core::{file_handlers::ExtensionHandler, App};
use std::{fs::File, io::Write, ops::Deref, path::PathBuf};

pub struct RomePath<'handler> {
	file: PathBuf,

	handler: Option<&'handler Box<dyn ExtensionHandler>>,
}

impl<'handler> Deref for RomePath<'handler> {
	type Target = PathBuf;

	fn deref(&self) -> &Self::Target {
		&self.file
	}
}

impl<'handler> RomePath<'handler> {
	pub fn new(path_to_file: &str) -> Self {
		Self {
			file: PathBuf::from(path_to_file),
			handler: None,
		}
	}

	/// Deduce the file handler based on the extension of the file.
	///
	/// Any error will default to the base file handler for now.
	///
	///
	/// ```rust
	/// use path::RomePath;
	/// use core::create_app;
	/// let app = create_app();
	/// let file = RomePath::new("file.js").deduce_handler(&app);
	/// let handler = file.get_handler();
	/// let expected = app.get_js_handler("js").unwrap();
	/// assert_eq!(handler.unwrap(), expected)
	/// ```
	pub fn deduce_handler(mut self, app: &'handler App) -> Self {
		if let None = self.extension() {
			return self;
		}
		let extension = self.extension().unwrap().to_str().unwrap();

		if let Some(handler) = app.get_handler(extension) {
			self.handler = Some(handler);
		}

		self
	}

	// TODO: handle error with diagnostic?
	/// Opens a file and returns a [File] in write mode
	pub fn open(&self) -> File {
		File::open(&self.file).expect("cannot open the file to format")
	}


	/// Accepts a file opened in read mode and saves into it
	pub fn save(&mut self, content: &str) -> Result<(), std::io::Error> {
		let mut file_to_write = File::create(&self.file).unwrap();
		file_to_write.write_all(content.as_bytes())
	}

	/// Returns the current handler associated to the file.
	///
	/// You need to call [deduce_handler] first in order to receive one. If not, [None] is always returned.
	pub fn get_handler(&self) -> Option<&Box<dyn ExtensionHandler>> {
		self.handler
	}
}

#[cfg(test)]
mod test {
	use crate::RomePath;
	use core::{
		create_app,
		file_handlers::{javascript::JsFileHandler, ExtensionHandler},
	};

	#[test]
	fn deduce_handler() {
		let app = create_app();
		let file = RomePath::new("file.js").deduce_handler(&app);
		let handler = file.get_handler();
		let expected = JsFileHandler {};
		assert_eq!(
			handler.unwrap().capabilities().format,
			expected.capabilities().format
		);
		assert_eq!(
			handler.unwrap().capabilities().lint,
			expected.capabilities().lint
		)
	}
}
