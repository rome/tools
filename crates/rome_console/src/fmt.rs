use std::{borrow::Cow, fmt, io, time::Duration};

use termcolor::{ColorSpec, WriteColor};

use crate::{markup, Markup, MarkupElement};

/// A stack-allocated linked-list of [MarkupElement] slices
enum MarkupElements<'a> {
    Root,
    Node(&'a Self, &'a [MarkupElement]),
}

impl<'a> MarkupElements<'a> {
    /// Iterates on all the element slices depth-first
    fn for_each(&self, func: &mut impl FnMut(&'a [MarkupElement])) {
        if let Self::Node(parent, elem) = self {
            parent.for_each(func);
            func(elem);
        }
    }
}

/// The [Formatter] is the `rome_console` equivalent to [std::fmt::Formatter]:
/// it's never constructed directly by consumers, and can only be used through
/// the mutable reference passed to implementations of the [Display] trait).
/// It manages the state of the markup to print, and implementations of
/// [Display] can call into its methods to append content into the current
/// printing session
pub struct Formatter<'fmt> {
    /// Stack of markup elements currently applied to the text being printed
    state: MarkupElements<'fmt>,
    /// Inner IO writer this [Formatter] will print text into
    writer: &'fmt mut dyn WriteColor,
}

impl<'fmt> Formatter<'fmt> {
    /// Create a new instance of the [Formatter] using the provided `writer` for printing
    pub(crate) fn new(writer: &'fmt mut dyn WriteColor) -> Self {
        Self {
            state: MarkupElements::Root,
            writer,
        }
    }

    /// Return a new instance of the [Formatter] with `elements` appended to its element stack
    fn with_elements<'b>(&'b mut self, elements: &'b [MarkupElement]) -> Formatter<'b> {
        Formatter {
            state: MarkupElements::Node(&self.state, elements),
            writer: self.writer,
        }
    }

    /// Write a piece of markup into this formatter
    pub fn write_markup(&mut self, markup: Markup) -> io::Result<()> {
        for node in markup.0 {
            let mut fmt = self.with_elements(node.elements);
            node.content.fmt(&mut fmt)?;
        }

        Ok(())
    }

    /// Applies the current format in `state` to `writer`, calls `func` to
    /// print a piece of text, then reset the printing format
    fn with_format(
        &mut self,
        func: impl FnOnce(&mut dyn WriteColor) -> io::Result<()>,
    ) -> io::Result<()> {
        let mut color = ColorSpec::new();
        self.state.for_each(&mut |elements| {
            for element in elements {
                element.update_color(&mut color);
            }
        });

        if let Err(err) = self.writer.set_color(&color) {
            self.writer.reset()?;
            return Err(err);
        }

        let result = func(self.writer);
        self.writer.reset()?;
        result
    }

    /// Write a slice of text into this formatter
    pub fn write_str(&mut self, content: &str) -> io::Result<()> {
        self.with_format(|writer| writer.write_all(content.as_bytes()))
    }

    /// Write formatted text into this formatter
    pub fn write_fmt(&mut self, content: fmt::Arguments) -> io::Result<()> {
        self.with_format(|writer| writer.write_fmt(content))
    }
}

/// Formatting trait for types to be displayed as markup, the `rome_console`
/// equivalent to [std::fmt::Display]
///
/// # Example
/// Implementing `Display` on a custom struct
/// ```
/// use std::io;
/// use rome_console::{fmt::{Display, Formatter}, markup};
///
/// struct Warning(String);
///
/// impl Display for Warning {
///     fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
///         fmt.write_markup(markup! {
///             <Warn>{self.0}</Warn>
///         })
///     }
/// }
///
/// let warning = Warning(String::from("content"));
/// markup! {
///     <Emphasis>{warning}</Emphasis>
/// };
/// ```
pub trait Display {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()>;
}

// Blanket implementations of Display for reference types
impl<'a, T> Display for &'a T
where
    T: Display + ?Sized,
{
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        T::fmt(self, fmt)
    }
}

impl<'a, T> Display for Cow<'a, T>
where
    T: Display + ToOwned + ?Sized,
{
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        T::fmt(self, fmt)
    }
}

// Simple implementations of Display calling through to write_str for types
// that implement Deref<str>
impl Display for str {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_str(self)
    }
}

impl Display for String {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_str(self)
    }
}

// Implement Display for Markup and Rust format Arguments
impl<'a> Display for Markup<'a> {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_markup(*self)
    }
}

impl<'a> Display for std::fmt::Arguments<'a> {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_fmt(*self)
    }
}

/// Implement [Display] for types that implement [std::fmt::Display] by calling
/// through to [Formatter::write_fmt]
macro_rules! impl_std_display {
    ($ty:ty) => {
        impl Display for $ty {
            fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
                write!(fmt, "{self}")
            }
        }
    };
}

impl_std_display!(char);
impl_std_display!(i8);
impl_std_display!(i16);
impl_std_display!(i32);
impl_std_display!(i64);
impl_std_display!(i128);
impl_std_display!(isize);
impl_std_display!(u8);
impl_std_display!(u16);
impl_std_display!(u32);
impl_std_display!(u64);
impl_std_display!(u128);
impl_std_display!(usize);

impl Display for Duration {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        use crate as rome_console;

        let secs = self.as_secs();
        if secs > 1 {
            return fmt.write_markup(markup! {
                {secs}<Dim>"s"</Dim>
            });
        }

        let millis = self.as_millis();
        if millis > 1 {
            return fmt.write_markup(markup! {
                {millis}<Dim>"ms"</Dim>
            });
        }

        let micros = self.as_micros();
        if micros > 1 {
            return fmt.write_markup(markup! {
                {micros}<Dim>"Âµs"</Dim>
            });
        }

        let nanos = self.as_nanos();
        fmt.write_markup(markup! {
            {nanos}<Dim>"ns"</Dim>
        })
    }
}
