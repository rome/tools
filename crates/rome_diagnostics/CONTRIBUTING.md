# Writing Diagnostics

The `rome_diagnostics` crate implements "Diagnostics", the generic concept of
how the various errors and issues that can be raised in the Rome codebase,
and displayed to the final user. This guide is aimed at
contributors to the Rome codebase to help writing diagnostics, both at the
technical level, and as general best practices to make diagnostics easier to
understand for the final user.

## What is a Diagnostic

Diagnostics are at the core of the experience of the Rome toolchain for the
users, and providing high quality diagnostics is crucial to making the usage of
Rome as frictionless as possible even when errors happen. What follows is a
list of best practice for writing good diagnostics:

- The Diagnostic should follow the [Technical Principles](https://rome.tools/#technical)
of the Rome project
- A diagnostic should not simply state that something went wrong, it should
	explain why it went wrong. Add explanations in log advices, and show hyperlinks
	to relevant documentation pages in case the user wants to know more.
- If possible, a diagnostic should also try to provide a way for the user to
	fix the issue. This can be in the form of a simple log advice, as a diff advice
	for a source code change, or as a command advice to prompt the user for a
	direct action. In any case, don't forget to add the `FIXABLE` tag to the
	diagnostic to highlight to the user it contains an actionable hint.
- Show don't tell: while log advices are highly versatile, always prefer
	showing a rich advice like a code frame, diff or command to a textual
	explanation if you can, since those are generally easier to understand and may
	provide additional context.


## The `Diagnostic` trait

At the lowest level, `Diagnostic` is a Rust trait implemented by various types
across the codebase that each represent a specific kind of diagnostic. All
these types have the following properties in common, as enforced by the trait:

- A `category`, the unique string of the error code associated with this
diagnostic type such as `lint/correctness/noArguments`, `args/invalid` or
`format/disabled`
- A `severity` defining how important the issue described by the diagnostic is
to the user. The severity of a diagnostic can be `Fatal`, `Error`, `Warning`,
`Information` or `Hint`
- A `description` text explaining the diagnostic to the user. This is what will
be displayed in contexts that do not support rich markup, such as embedded
error popovers in editors. Since the description is shown in contexts where the
advices cannot be rendered, it should provide as much information as possible about the
error.
- A `message`, a piece of rich markup that will be displayed at the top of the
diagnostic advices. Contrary to the description, the message is displayed along
with the other advices of the diagnostic and should only provide a short and
clear explanation of the error, while advices will be used to provide
additional information with the appropriate context.
- A set of `advices`, the main building blocks used to provide the user with
rich information about the diagnostic. Depending on the use case, advices can be
used to log pieces of markup, lists of items, annotated code frames, text diffs,
Rust backtraces, command lines, or groups of more advices.
- An optional set of `verbose_advices` the user can optionally enable (at the
moment these are always printed but they will eventually have to be enabled,
for instance by passing a `--verbose` flag to the CLI). Verbose advices are
printed in a separate section below the main advices and can be used to provide
further information about a given diagnostic (for instance a fixable linter
diagnostic may print a verbose advice instructing the user to run
`rome check --apply` to automatically apply the fix).
- A `location` describing where the error happened. It can be a path to a file
on the file system, a command line argument, or an arbitrary memory buffer. It may
optionally specify a specific range within the text content of this resource,
as well as embed said text content to faciliate it's retrieval when displaying
code frames in the diagnostic. Conceptually the `location` points to a
"resource" this error originated from.
- `tags` conveying additional informations about the diagnostic: if the
diagnostic has information on how it can be fixed, if it resulted from an
internal error in Rome and was not directly caused by the user, if it is being
emitted to warn the user about unused or deprecated code.
- An optional `source`, another diagnostic that details the low-level reason
why this diagnostic was emitted: for instance a diagnostic reporting a failed
request to a remote server may have a deserialization error for the server
response as its cause. Conceptually the `source` points to a `Diagnostic` this
error originated from, so while they are similar in purpose (explaining where
an error comes from to the user) the `location` and `source` properties are not
incompatible: taking from the above example of a "failed request" diagnostic,
it may have a "deserialization error" diagnostic as a `source`, and the body
of the invalid response as its `location`.

## How to implement Diagnostic

In theory a diagnostic is created by implementing the `Diagnostic` trait on a
type. In practice since there's a lot of methods to implement, the
`rome_diagnostics` crate exposes a `derive` procedural macro to make
implementing the trait easier:

```rust
// The Diagnostic trait requires Debug to be implemented
#[derive(Debug, Diagnostic)]
// The category, severity, description, message, location and tags can be
// specified statically on the type itself using the #[diagnostic] attribute
#[diagnostic(severity = Warning, category = "internalError/fs")]
struct UnhandledDiagnostic {
    // All the diagnostic properties can also be derived from fields of the
    // struct using the corresponding attribute
    // A single field may have multiple attributes, however most attributes can
    // only be specified once either statically on the whole struct or on a
    // single field. The only exception to this is #[advice] (and
    // #[verbose_advice]), since all advices will be recorded into the
    // diagnostic in the same order they are declared in the struct
    #[message]
    #[description]
    #[advice]
    file_kind: UnhandledKind
}
```

This should be enough to define most of the properties of a diagnostic, but
advices are a bit more complex. Fields that have the `#[advice]` or
`#[verbose_advice]` attribute are expected to implement the `Advices` trait.
This trait allows arbitrary types to record advices on the diagnostic that
contains them:

```rust
impl Advices for UnhandledKind {
    fn record(&self, visitor: &mut dyn Visit) -> io::Result<()> {
        visitor.record_log(LogCategory::Info, self.advice_message())
    }
}
```

In order to add advices to a diagnostic, you can either use some of the helper
types in the `rome_diagnostics::v2` module (`CodeFrameAdvice`, `CommandAdvice`,
`DiffAdvice`, `LogAdvice`), or implement the `Advices` trait yourself if you
want more control over how the advices are recorded (this can be useful to
optimize the performance of a diagnostic by reducing how much memory it
allocates).

The category may also require some special care if you're declaring a new one,
since all diagnostic categories have to be statically registered you'll need to
add it to `crates/rome_diagnostics_categories/src/categories.rs`

