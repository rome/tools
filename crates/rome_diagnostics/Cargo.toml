[package]
name = "rome_diagnostics"
version = "0.0.1"
authors = { workspace = true }
edition = { workspace = true }
license = { workspace = true }
description = "Pretty error reporting library"
documentation = "https://rustdocs.rome.tools/rome_diagnostics/index.html"
categories = ["development-tools"]

[[example]]
name = "cli"
test = true

[[example]]
name = "fs"
test = true

[[example]]
name = "lint"
test = true

[[example]]
name = "serde"
test = true

[dependencies]
rome_rowan = { version = "0.0.1", path = "../rome_rowan" }
rome_console = { version = "0.0.1", path = "../rome_console", features = ["serde_markup"] }
rome_diagnostics_macros = { version = "0.0.1", path = "../rome_diagnostics_macros" }
rome_diagnostics_categories = { version = "0.0.1", path = "../rome_diagnostics_categories", features = ["serde"] }
rome_text_edit = { version = "0.0.1", path = "../rome_text_edit" }
rome_text_size = { version = "0.0.1", path = "../rome_text_size" }
unicode-width = "0.1.9"
serde = { version = "1.0.133", features = ["derive"] }
termcolor = "1.1.2"
schemars = { version = "0.8.10", optional = true }
bitflags = "1.3.2"
backtrace = "0.3.66"
pico-args = "0.5.0"

[features]
schema = ["schemars", "rome_text_edit/schemars", "rome_diagnostics_categories/schemars"]

[dev-dependencies]
trybuild = "1.0"
serde_json = "1.0.74"
