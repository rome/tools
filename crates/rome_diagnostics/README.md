# `rome_diagnostics`

This crate contains the types and utility functions used to implement errors
and diagnostics in the Rome codebase

## Acknowledgement

This crate was initially forked from [rslint_errors](https://github.com/rslint/rslint/tree/master/crates/rslint_errors). The design of the new `Diagnostic` trait, `Error` struct, `Context` trait, and the `Diagnostic` derive macro in `rome_diagnostics_macros` are inspired by various fantastic crates in the Rust error handling space: [miette](https://github.com/zkat/miette), [anyhow](https://github.com/dtolnay/anyhow) and [thiserror](https://github.com/dtolnay/thiserror)
