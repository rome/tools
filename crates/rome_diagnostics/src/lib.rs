#![deny(rust_2018_idioms)]

pub mod file;

mod diagnostic;
mod emit;
mod suggestion;

pub use diagnostic::{Diagnostic, Footer, SubDiagnostic};
pub use emit::{DiagnosticHeader, Emitter};
pub use file::Span;
pub use suggestion::*;

pub use rome_console::codespan::Severity;
pub use termcolor;

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum DiagnosticTag {
    Unnecessary,
    Deprecated,
    Both,
}

impl DiagnosticTag {
    pub fn is_unecessary(&self) -> bool {
        matches!(self, DiagnosticTag::Unnecessary | DiagnosticTag::Both)
    }

    pub fn is_deprecated(&self) -> bool {
        matches!(self, DiagnosticTag::Deprecated | DiagnosticTag::Both)
    }
}

/// Indicicates how a tool should manage this suggestion.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum Applicability {
    /// The suggestion is definitely what the user intended.
    /// This suggestion should be automatically applied.
    Always,
    /// The suggestion may be what the user intended, but it is uncertain.
    /// The suggestion should result in valid JavaScript/TypeScript code if it is applied.
    MaybeIncorrect,
    /// The suggestion contains placeholders like `(...)` or `{ /* fields */ }`.
    /// The suggestion cannot be applied automatically because it will not result in valid JavaScript/TypeScript code.
    /// The user will need to fill in the placeholders.
    HasPlaceholders,
    /// The applicability of the suggestion is unknown.
    Unspecified,
}

pub const MAXIMUM_DISPLAYABLE_DIAGNOSTICS: u8 = 50;
