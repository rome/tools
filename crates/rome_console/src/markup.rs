use std::{
    borrow::Cow,
    fmt::{self, Debug},
    io,
};

use rome_text_size::TextSize;
use termcolor::{Color, ColorSpec};

use crate::fmt::{Display, Formatter, MarkupElements, Write};

/// Enumeration of all the supported markup elements
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
#[cfg_attr(
    feature = "serde",
    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
)]
pub enum MarkupElement<'fmt> {
    Emphasis,
    Dim,
    Italic,
    Underline,
    Error,
    Success,
    Warn,
    Info,
    Hyperlink { href: Cow<'fmt, str> },
}

impl<'fmt> MarkupElement<'fmt> {
    /// Mutate a [ColorSpec] object in place to apply this element's associated
    /// style to it
    pub(crate) fn update_color(&self, color: &mut ColorSpec) {
        match self {
            // Text Styles
            MarkupElement::Emphasis => {
                color.set_bold(true);
            }
            MarkupElement::Dim => {
                color.set_dimmed(true);
            }
            MarkupElement::Italic => {
                color.set_italic(true);
            }
            MarkupElement::Underline => {
                color.set_underline(true);
            }

            // Text Colors
            MarkupElement::Error => {
                color.set_fg(Some(Color::Red));
            }
            MarkupElement::Success => {
                color.set_fg(Some(Color::Green));
            }
            MarkupElement::Warn => {
                color.set_fg(Some(Color::Yellow));
            }
            MarkupElement::Info => {
                // Blue is really difficult to see on the standard windows command line
                #[cfg(windows)]
                const BLUE: Color = Color::Cyan;
                #[cfg(not(windows))]
                const BLUE: Color = Color::Blue;

                color.set_fg(Some(BLUE));
            }

            MarkupElement::Hyperlink { .. } => {}
        }
    }

    fn to_owned(&self) -> MarkupElement<'static> {
        match self {
            MarkupElement::Emphasis => MarkupElement::Emphasis,
            MarkupElement::Dim => MarkupElement::Dim,
            MarkupElement::Italic => MarkupElement::Italic,
            MarkupElement::Underline => MarkupElement::Underline,
            MarkupElement::Error => MarkupElement::Error,
            MarkupElement::Success => MarkupElement::Success,
            MarkupElement::Warn => MarkupElement::Warn,
            MarkupElement::Info => MarkupElement::Info,
            MarkupElement::Hyperlink { href } => MarkupElement::Hyperlink {
                href: Cow::Owned(match href {
                    Cow::Borrowed(href) => href.to_string(),
                    Cow::Owned(href) => href.clone(),
                }),
            },
        }
    }
}

/// Implementation of a single "markup node": a piece of text with a number of
/// associated styles applied to it
#[derive(Copy, Clone)]
pub struct MarkupNode<'fmt> {
    pub elements: &'fmt [MarkupElement<'fmt>],
    pub content: &'fmt dyn Display,
}

#[derive(Clone, PartialEq, Eq, Hash)]
#[cfg_attr(
    feature = "serde",
    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
)]
pub struct MarkupNodeBuf {
    pub elements: Vec<MarkupElement<'static>>,
    pub content: String,
}

impl Debug for MarkupNodeBuf {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        for element in &self.elements {
            write!(fmt, "<{element:?}>")?;
        }
        write!(fmt, "{:?}", self.content)?;
        for element in self.elements.iter().rev() {
            write!(fmt, "</{element:?}>")?;
        }
        if fmt.alternate() && self.content.contains('\n') {
            writeln!(fmt)?;
        }
        Ok(())
    }
}

/// Root type returned by the `markup` macro: this is simply a container for a
/// list of markup nodes
///
/// Text nodes are formatted lazily by storing an [fmt::Arguments] struct, this
/// means [Markup] shares the same restriction as the values returned by
/// [format_args] and can't be stored in a `let` binding for instance
#[derive(Copy, Clone)]
pub struct Markup<'fmt>(pub &'fmt [MarkupNode<'fmt>]);

impl<'fmt> Markup<'fmt> {
    pub fn to_owned(&self) -> MarkupBuf {
        let mut result = MarkupBuf(Vec::new());
        // SAFETY: The implementation of Write for MarkupBuf below always returns Ok
        Formatter::new(&mut result).write_markup(*self).unwrap();
        result
    }
}

#[derive(Clone, Default, PartialEq, Eq, Hash)]
#[cfg_attr(
    feature = "serde",
    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
)]
pub struct MarkupBuf(pub Vec<MarkupNodeBuf>);

impl MarkupBuf {
    pub fn is_empty(&self) -> bool {
        self.0.iter().all(|node| node.content.is_empty())
    }

    pub fn len(&self) -> TextSize {
        self.0.iter().map(|node| TextSize::of(&node.content)).sum()
    }
}

impl Write for MarkupBuf {
    fn write_str(&mut self, elements: &MarkupElements, content: &str) -> io::Result<()> {
        let mut styles = Vec::new();
        elements.for_each(&mut |elements| {
            styles.extend(elements.iter().map(MarkupElement::to_owned));
            Ok(())
        })?;

        if let Some(last) = self.0.last_mut() {
            if last.elements == styles {
                last.content.push_str(content);
                return Ok(());
            }
        }

        self.0.push(MarkupNodeBuf {
            elements: styles,
            content: content.into(),
        });

        Ok(())
    }

    fn write_fmt(&mut self, elements: &MarkupElements, content: fmt::Arguments) -> io::Result<()> {
        let mut styles = Vec::new();
        elements.for_each(&mut |elements| {
            styles.extend(elements.iter().map(MarkupElement::to_owned));
            Ok(())
        })?;

        if let Some(last) = self.0.last_mut() {
            if last.elements == styles {
                last.content.push_str(&content.to_string());
                return Ok(());
            }
        }

        self.0.push(MarkupNodeBuf {
            elements: styles,
            content: content.to_string(),
        });
        Ok(())
    }
}

impl Display for MarkupBuf {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        let nodes: Vec<_> = self
            .0
            .iter()
            .map(|node| MarkupNode {
                elements: &node.elements,
                content: &node.content,
            })
            .collect();

        fmt.write_markup(Markup(&nodes))
    }
}

impl Debug for MarkupBuf {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        for node in &self.0 {
            write!(fmt, "{node:?}")?;
        }
        Ok(())
    }
}
