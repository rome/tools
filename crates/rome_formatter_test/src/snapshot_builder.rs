use rome_diagnostics::console::fmt::{Formatter, Termcolor};
use rome_diagnostics::console::markup;
use rome_diagnostics::PrintDiagnostic;
use rome_diagnostics::{termcolor, DiagnosticExt};
use rome_formatter::Printed;
use rome_parser::AnyParse;
use std::ffi::OsStr;
use std::fmt;
use std::fmt::Write;
use std::path::Path;

#[derive(serde::Serialize)]
struct TestInfo {
    test_file: String,
}

pub struct SnapshotOutput<'a> {
    content: &'a str,
    index: Option<usize>,
}

impl<'a> SnapshotOutput<'a> {
    pub fn new(content: &'a str) -> Self {
        SnapshotOutput {
            content,
            index: None,
        }
    }

    pub fn with_index(mut self, index: usize) -> Self {
        self.index = Some(index);
        self
    }
}

pub struct SnapshotBuilder<'a> {
    input_file: &'a Path,
    snapshot: String,
}

impl<'a> SnapshotBuilder<'a> {
    pub fn new(input_file: &'a Path) -> Self {
        SnapshotBuilder {
            input_file,
            snapshot: String::new(),
        }
    }

    pub fn with_input(mut self, input: &str) -> Self {
        writeln!(self.snapshot).unwrap();
        writeln!(self.snapshot, "# Input").unwrap();
        writeln!(self.snapshot).unwrap();
        self.write_extension();
        self.snapshot.push_str(input);
        writeln!(self.snapshot).unwrap();
        writeln!(self.snapshot, "```").unwrap();
        writeln!(self.snapshot).unwrap();
        writeln!(self.snapshot).unwrap();

        self
    }

    pub fn with_separator(mut self) -> Self {
        writeln!(self.snapshot, "=============================").unwrap();
        writeln!(self.snapshot).unwrap();

        self
    }

    pub fn with_prettier_diff(mut self, prettier_diff: &str) -> Self {
        writeln!(self.snapshot, "# Prettier differences").unwrap();
        writeln!(self.snapshot).unwrap();
        writeln!(self.snapshot, "```diff").unwrap();
        self.snapshot.push_str(prettier_diff);
        writeln!(self.snapshot, "```").unwrap();
        writeln!(self.snapshot).unwrap();

        self
    }

    pub fn with_multiple_outputs(mut self) -> Self {
        writeln!(self.snapshot, "# Outputs").unwrap();
        writeln!(self.snapshot).unwrap();

        self
    }

    pub fn with_output_and_options<T>(mut self, output: SnapshotOutput, options: T) -> Self
    where
        T: fmt::Display,
    {
        self.write_output_header(&output);

        writeln!(self.snapshot).unwrap();

        writeln!(self.snapshot, "-----").unwrap();
        write!(self.snapshot, "{}", options).unwrap();
        writeln!(self.snapshot, "-----").unwrap();
        writeln!(self.snapshot).unwrap();

        self.write_extension();
        self.snapshot.push_str(output.content);
        writeln!(self.snapshot, "```").unwrap();
        writeln!(self.snapshot).unwrap();

        self
    }

    pub fn with_output(mut self, output: SnapshotOutput) -> Self {
        self.write_output_header(&output);

        writeln!(self.snapshot).unwrap();

        self.write_extension();
        self.snapshot.push_str(output.content);
        writeln!(self.snapshot, "```").unwrap();
        writeln!(self.snapshot).unwrap();

        self
    }

    pub fn with_unimplemented(mut self, formatted: &Printed) -> Self {
        if !formatted.verbatim_ranges().is_empty() {
            writeln!(self.snapshot).unwrap();
            writeln!(self.snapshot).unwrap();
            self.snapshot.push_str("## Unimplemented nodes/tokens");
            writeln!(self.snapshot).unwrap();
            writeln!(self.snapshot).unwrap();

            for (range, text) in formatted.verbatim() {
                writeln!(self.snapshot, "{:?} => {:?}", text, range).unwrap();
            }
        }

        self
    }

    pub fn with_errors(mut self, parsed: &AnyParse, parse_input: &str) -> Self {
        if !parsed.has_errors() {
            return self;
        }

        let file_name = self.input_file.file_name().and_then(OsStr::to_str).unwrap();

        let mut buffer = termcolor::Buffer::no_color();

        for diagnostic in parsed.diagnostics() {
            let error = diagnostic
                .clone()
                .with_file_path(file_name)
                .with_file_source_code(parse_input);
            Formatter::new(&mut Termcolor(&mut buffer))
                .write_markup(markup! {
                    {PrintDiagnostic::verbose(&error)}
                })
                .expect("failed to emit diagnostic");
        }

        writeln!(self.snapshot, "# Errors").unwrap();
        writeln!(self.snapshot, "```").unwrap();
        writeln!(
            self.snapshot,
            "{}",
            std::str::from_utf8(buffer.as_slice()).expect("non utf8 in error buffer")
        )
        .unwrap();
        writeln!(self.snapshot, "```").unwrap();
        writeln!(self.snapshot).unwrap();

        self
    }

    pub fn with_lines_exceeding_max_width(mut self, output: &str, max_width: usize) -> Self {
        let mut lines_exceeding_max_width = output
            .lines()
            .enumerate()
            .filter(|(_, line)| line.len() > max_width)
            .peekable();

        if lines_exceeding_max_width.peek().is_some() {
            writeln!(
                self.snapshot,
                "# Lines exceeding max width of {max_width} characters"
            )
            .unwrap();
            writeln!(self.snapshot, "```").unwrap();

            for (index, line) in lines_exceeding_max_width {
                let line_number = index + 1;
                writeln!(self.snapshot, "{line_number:>5}: {line}").unwrap();
            }
            writeln!(self.snapshot, "```").unwrap();
            writeln!(self.snapshot).unwrap();
        }

        self
    }

    pub fn finish(self, relative_file_name: &str) {
        let file_name = self.input_file.file_name().and_then(OsStr::to_str).unwrap();

        let info = TestInfo {
            test_file: relative_file_name.to_owned(),
        };

        insta::with_settings!({
            prepend_module_to_snapshot => false,
            snapshot_path => self.input_file.parent().unwrap(),
            omit_expression => true,
            raw_info => &info.test_file.into()
        }, {
            insta::assert_snapshot!(file_name, self.snapshot);
        });
    }
}

impl SnapshotBuilder<'_> {
    fn write_extension(&mut self) {
        let file_extension = self.input_file.extension().unwrap().to_str().unwrap();
        writeln!(self.snapshot, "```{file_extension}").unwrap();
    }

    fn write_output_header(&mut self, output: &SnapshotOutput) {
        if let Some(index) = output.index {
            writeln!(self.snapshot, "## Output {index}").unwrap();
        } else {
            writeln!(self.snapshot, "# Output").unwrap();
        }
    }
}
