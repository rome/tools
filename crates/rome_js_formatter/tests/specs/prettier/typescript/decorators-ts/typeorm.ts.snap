---
source: crates/rome_formatter_test/src/snapshot_builder.rs
info: typescript/decorators-ts/typeorm.ts
---

# Input

```ts
@Entity()
export class Board {

    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    slug: string;

    @Column()
    name: string;

    @Column()
    theme: string;

    @Column()
    description: string;

    @OneToMany(type => Topic, topic => topic.board)
    topics: Topic[]

}

```


# Prettier differences

```diff
--- Prettier
+++ Rome
@@ -15,6 +15,9 @@
   @Column()
   description: string;
 
-  @OneToMany((type) => Topic, (topic) => topic.board)
+  @OneToMany(
+    (type) => Topic,
+    (topic) => topic.board,
+  )
   topics: Topic[];
 }
```

# Output

```ts
@Entity()
export class Board {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  slug: string;

  @Column()
  name: string;

  @Column()
  theme: string;

  @Column()
  description: string;

  @OneToMany(
    (type) => Topic,
    (topic) => topic.board,
  )
  topics: Topic[];
}
```


